scrollable_overflow implementation

wlott opened the pull request:
# Objective

Replaces `content_size: Size<f32>` with `scrollable_overflow: Rect<f32>` and tries to match the definition of "scrollable overflow rectangle" in [CSS Overflow 3, 2.2. Scrollable Overflow](https://www.w3.org/TR/css-overflow-3/#scrollable).

Fixes #871 

- [ ] **Changes that will affect external library users must update RELEASES.md before they will be merged.**

## Context

See #871.  

## Feedback wanted

I added several helper/utility `Rect` methods to geometry.rs.  I wasn't sure if they are of sufficient generality to be methods directly on `Rect` or whether they should be helper functions in some utility module.  Also, if they are `Rect` methods, I wasn't sure if they should be `pub` or just `pub(crate)`.  I saw examples of both, so I couldn't just do the same as the others.

I made some changes to `gentest/src/main.rs` (which unfortunately means that every generated file shows up in this commit) but I don't think they should be controversial:

- I pulled long string literals for leaf nodes out onto their own line to make rustfmt not throw up its hands in despair.  
- I changed the assertion logic to check every node then assert once instead of asserting at each comparison.  This way you can see all the things you got wrong, not just the first.
- And I (tentatively) turned on the `scroll_width`/`scroll_height` checks for all nodes instead of limiting them to scroll containers.  Hopefully I'll be able to fix the last few fails and won't have to turn that back off.

The actual logic for the scrollable overflow computations follows a very similar structure to the content_size computations that I replaced.  There were some subtleties that took some trial-and-error to figure out, like the fact that the scrollable overflow from inflow child elements is expanded by the container's padding but the scrollable overflow from out-of-flow elements is taken as is.  But what it is doing is pretty straight forward.  The most non-obvious bit I think is that `LayoutOutput` only contains some of the scrollable overflow and it is up to the caller of `tree.perform_child_layout` to fill in the rest before calling `tree.set_unrounded_layout`.  The `compute_*_layout` functions just don't have enough information to compute it all in one place.

And finally, I'm sure this stuff needs more test cases to be fully exercised, but I just don't know enough CSS to figure out all that might entail.  I know a million times more about CSS layout now than I did when I started this work, but, well, that's like saying 10^-6 is a million times more than 10^-12: both are still small numbers.

Commit:
scrollable_overflow WIP snapshot

wlott commented:
Oh, something else I forgot to mention.  By the time I got done with this work, the text "content_size" shows up pretty much only in `#[cfg(feature = "content_size")]` annotations.  It really isn't the "content_size" feature any more.  Couple of options:

- We could add a `content_size` field back into `Layout` that tries to duplicate the prior semantics.
- We could change all my scrollable overflow stuff to gate on a new feature, say `scrollable_overflow`.
- We could do that but also put back all the `content_size` stuff so people could choose between the two feature sets.
- `scrollable_overflow` feature with just the new stuff then a `content_size` feature that requires `scrollable_overflow` and just adds a `content_size` field that tries to duplicate prior semantics.

I guess it boils down to separate questions:
- do we add a new scrollable_overflow feature?
- do we remove the old content_size feature?
- do we preserve the old `content_size` field in `Layout` (and possibly `LayoutOutput`)
- if there is a `content_size` field, do we try to compute it from the scrollable overflow or do we preserve both sets of code under different feature gates?

My personal bias is towards renaming the feature and pretending content_size never existed.  But I come out of a career of smallish game shops where there were no external clients for the code so reducing maintenance burden was much more valuable than backwards compatibility.

nicoburns commented on scripts/gentest/src/main.rs:
```diff
@@ -270,44 +272,50 @@ fn generate_assertions(ident: &str, node: &Value, use_rounding: bool) -> TokenSt
+                eprintln!("{:?}.{} mismatch: expected {} actual {}", #ident, stringify!(#expr), #expected, layout.#expr);
+            }
+        )
+    };
```
This is very nice! At some point I am hoping to be able to print an actual diff between the expected and actual trees. But this is a much easier way of getting some of the benefit of that.

nicoburns commented on src/compute/common/content_size.rs:
```diff
@@ -1,32 +1,84 @@
+    // - what does this mean?
+    // - is this predicated the same as the previous item?
+    // - assuming the margins are positive, don't this subsume the
+    //   previous item?
```
Are you missing margin-box vs border-box here? This seems to be saying that margins should extend the overflow scroll contribution for flex and grid items. Whereas they would be ignored for other kinds of box.

nicoburns reviewed (commented):
This is overall looking good. I've left some comments.

We should compare benchmark results against main. To run benchmarks: `cargo bench -p taffy_benchmarks`.

I haven't dug into the failing tests yet.

wlott commented on scripts/gentest/src/main.rs:
```diff
@@ -935,7 +954,20 @@ fn generate_node_context(text_content: &str, writing_mode: Option<&str>, aspect_
+    // If the string literal is too long, rustfmt gives up on the
+    // whole line.  So pull long string literals out onto their own
+    // line so that the Style constructor still gets formatted.
+    if string_lit.to_string().len() > 100 {
```
The quote->format->length bit is because rustfmt gives up when the source representation of the string literal is too long, not when the string contents are too long.  There were strings like `"HH\u{200b}HH\u{200b}HH..."` that overflowed rustfmt's cutoff but still had a low character count.

Another option would be to use the [prettyplease](https://crates.io/crates/prettyplease) crate which appears to have been purpose-built to handle good-enough formatting for generated token streams.

wlott commented on src/tree/layout.rs:
```diff
@@ -195,28 +199,35 @@ impl LayoutOutput {
     pub fn from_sizes_and_baselines(
         size: Size<f32>,
-        #[cfg_attr(not(feature = "content_size"), allow(unused_variables))] content_size: Size<f32>,
+        #[cfg(feature = "content_size")] descendent_scrollable_overflow: Rect<f32>,
```
Ah.  I had not realized that was a possibility.  How about I this field (and `scrollable_overflow` in `Layout`) always contain `Rect::new_empty()` when the feature isn't enabled?

wlott commented on src/tree/layout.rs:
```diff
@@ -324,24 +337,34 @@ impl Layout {
+    /// Return the scroll height of the node.  The scroll height is
+    /// the difference between the padding box height and the portion
+    /// of the scrollable overflow below the top padding edge.  The
+    /// scroll origin is currently always the padding box's top-left
```
Good point.

nicoburns commented on src/compute/flexbox.rs:
```diff
@@ -364,13 +373,56 @@ fn compute_preliminary(tree: &mut impl LayoutFlexboxContainer, node: NodeId, inp
+    // so we do it here before adding the out-of-flow (absolute)
+    // children.
+    if constants.is_scroll_container {
+        scrollable_overflow = scrollable_overflow.inset_by(-constants.padding);
```
> As for position: relative, this is one of those areas where my lack of familiarity with CSS makes it hard for me to intuit things like that

This is not intuitive at all! I also missed this at my first pass of this feature and I have almost 20 years of experience with CSS.

To be clear, I think the logic for `position: relative` nodes is that you need the union of:
- The "static position" (position before it is modified by relative positioning) + padding
- The "inset-modified position" (position after it is modified by relative positioning).

I also don't think we should be scanning the list twice, so either option 2 or 3 works.

(leaving it for a follow-up is still fine)

wlott commented on src/compute/common/content_size.rs:
```diff
@@ -1,32 +1,84 @@
+    //     context. [CSS3-TRANSFORMS]
+    if item_rect.right < container_scroll_origin.x || item_rect.bottom < container_scroll_origin.y {
+        return;
+    }
```
I'm pretty sure this test is needed from some of the fixtures to match chrome's results.  But I agree, it does diminish the utility of scrollable overflow for hit testing.  It is unfortunate that they stuck this wart in the spec just to accommodate legacy web pages that were abusing negative positions.

Commit:
gentest: explicitly set overflow instead relying on defaults.

Commit:
Easy fixes from first round of feedback.

Added `Rect::outset_by` and changed `inset_by(-something)` to use it
instead of `-`.

Made `Layout` and `LayoutOutput` include the overflow fields even when
the content_size feature isn't enabled.  I'm not 100% happy with the
`..default`s I added but it was either that or add lots of
`#[cfg(not(feature="content_size"))]`.  I went with the use of
`..default` as the lesser of two evils.

nicoburns commented:
Looking into the `justify_content_row_max_width_and_margin` test. Taffy's result seems to agree with Firefox but not Chrome here. Additionally, Chrome starts giving this results if overflow is set to hidden/scroll/auto.
